### PR TITLE
[Hardware][NVIDIA] Add non-NVML CUDA mode for Jetson

### LINKED ISSUE #9728: [Bug]:  Jetson support regression
### Your current environment

see comment

### Model Input Dumps

_No response_

### 🐛 Describe the bug

I'm currently in the process of updating the [nixpkgs build definition](https://github.com/conroy-cheers/nixpkgs/blob/update-vllm-0.6.3/pkgs/development/python-modules/vllm/default.nix) for vLLM from 0.5.3.post1 to 0.6.3.post1.

Build works great on x86_64 machine with NVIDIA GPU, and have successfully built for aarch64 Jetson Orin AGX with correct dependency / CUDA library versions. Unfortunately, at runtime on the Jetson, vLLM fails to start engine process:
```
  File "/nix/store/x5bz4k7prp7vdvm4i01y0z0lbi4awad7-python3.12-vllm-0.6.3.post1/lib/python3.12/site-packages/vllm/utils.py", line 799, in current_memory_usage
    return mem
           ^^^
UnboundLocalError: cannot access local variable 'mem' where it is not associated with a value
```
Have traced this back to the root cause - [newer versions of vLLM use NVML to detect CUDA support](https://github.com/vllm-project/vllm/blob/v0.6.3.post1/vllm/platforms/cuda.py).
NVML is not supported on Jetson, so unfortunately it isn't possible to proceed without making changes to vLLM here.

I'd be happy to open a PR to add a fallback mode on CUDA platforms where NVML isn't supported, if the maintainers would be open to it? Otherwise I will have to maintain a patch downstream.

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by conroy-cheers**
NVIDIA Jetson devices do not support the NVML protocol.

This PR adds a discrete `CudaJetsonPlatform` platform which does not rely on NVML, and is automatically used
on Jetson devices when NVML cannot be initialized. The Jetson platform supports all CUDA functionality, but does
not support NVLink; platform-specific checks are adjusted where appropriate to account for this.

Jetson Orin devices use CUDA capability 8.7 which is mostly identical to 8.6; capability 8.7 has been added to definitions
where 8.6 is used.

This has been built and tested on Jetson Orin AGX.

Fixes #9728

### CODE REVIEW

**Comment by youkaichao** [comment]
I don't think you need to create another platform class.

if Jetson does not support nvlink, you can just detect Jetson and return false for the nvlink detection.

**Comment by youkaichao** [comment]
creating a platform just for Jetson is too overkill.

**Comment by conroy-cheers** [comment]
It's not just NVLink; currently that's only used for custom allreduce as far as I can tell.

The CUDA platform class assumes NVML is available throughout to avoid initialising a CUDA context, but Jetson doesn't have NVML, so a different approach needs to be taken.

I didn't want to pollute the CUDA platform by initializing a CUDA context, so created a different one.

Happy to rework if you prefer a different approach?

**Comment by conroy-cheers** [comment]
Alternatively, could try to abstract away the NVML calls to make calls to an alternative backend on Jetson - [jetson_stats](https://github.com/rbonghi/jetson_stats) might be usable,  but not sure if all features will be implemented (especially if we want to use other NVML features in future).

**Comment by conroy-cheers** [comment]
I believe Jetson is also currently the only CUDA-supporting device to use a unified memory configuration, so some optimisations may be possible there (e.g. handover of weights between CPU and GPU without copy).

However it could be argued to save the separate backend for when such optimisations are actually implemented.

**Comment by tlrmchlsmth** [comment]
Looking at the logs, adding 8.7 to the ARCHS is not affecting the binary size at all:

This PR:
```
[2024-10-27T13:36:44Z] #28 0.789 Wheel dist/vllm-0.6.3.post2.dev123+gda79e3e5.cu124-cp38-abi3-linux_x86_64.whl is within the allowed size (184.58 MB).
```

Latest main:
```
[2024-10-28T05:09:01Z] #28 0.878 Wheel dist/vllm-0.6.3.post2.dev125+g32176fee.cu124-cp38-abi3-linux_x86_64.whl is within the allowed size (184.58 MB).
```

Does this make sense?

**Comment by tlrmchlsmth** [comment]
> Possibly - as far as I know 8.6 and 8.7 are essentially compatible, so it's possible it's just the same kernels being linked twice?
> 
> I did have to add the 8.7 capability to get vLLM running on Orin though; without, it'd complain that there was no compatible kernel available.

Ok, I didn't catch this before, but this PR isn't adding the sm87 kernels to the wheel file (just adds support for compiling them). IIUC, Jetson is ARM-only and the vLLM wheels are only built for x86, so all that looks good to me for this PR

**Review by tlrmchlsmth — COMMENTED** [review]
I leave it to @youkaichao to decide whether it makes sense to have Jetson in its own platform, but lgtm otherwise

**Comment by youkaichao** [comment]
I can find https://github.com/dusty-nv/jetson-containers/tree/master/packages/llm/vllm who builds vllm on jetson. I don't see any manipulation like this.

**Comment by conroy-cheers** [comment]
> I can find https://github.com/dusty-nv/jetson-containers/tree/master/packages/llm/vllm who builds vllm on jetson. I don't see any manipulation like this.

Just investigated this. NVIDIA JetPack 6 includes NVML support, so the alternate backend will not be needed on this platform.

However, JetPack 5 and older still do not support NVML, and JetPack 6 is only releasing for the Orin series, not any older Jetson devices. So if we want to support JetPack 5 devices we will need to use the alternate non-NVML approach.

**Comment by youkaichao** [comment]
you can add an if-else branch for all functions inside `cuda.py` , to deal with non-nvml case. I'm fine with that.

however, I don't think you should touch anything outside of `cuda.py` . that's the whole point of introducing platform abstraction.

**Comment by conroy-cheers** [comment]
Cool, I will refactor to remove the CUDA_JETSON platform and roll it into `cuda.py`.

**Comment by conroy-cheers** [comment]
Refactor done; have retested on Jetson; ready for re-review.

File: vllm/platforms/cuda.py
Comment (youkaichao) [inline vllm/platforms/cuda.py:130]: can you add another class like `NonNVMLContext` and implement it there? don't put too many `if-else` here.

**Comment by conroy-cheers** [comment]
Bump; need re-review

**Comment by conroy-cheers** [comment]
@youkaichao
Sorry for the ping, would you be able to re-review this? It would be great to be able to get this merged soon.
Thank you

**Comment by cthulhu-tww** [comment]
> @youkaichao 抱歉，您能重新审核一下吗？如果能尽快合并就太好了。 谢谢

Can this branch build VLLM directly without relying on Jetson containers?

**Comment by youkaichao** [comment]
will review recently.

**Comment by conroy-cheers** [comment]
> > @youkaichao 抱歉，您能重新审核一下吗？如果能尽快合并就太好了。 谢谢
> 
> 
> 
> Can this branch build VLLM directly without relying on Jetson containers?

Yes, I am currently building this directly on Jetson outside of containers. Build process is fully specified by [my nixpkgs branch](https://github.com/conroy-cheers/nixpkgs/blob/update-vllm-jetson/pkgs/development/python-modules/vllm/default.nix); for Jetson support make sure to override the source hash to use this branch.

You are welcome to build it directly using Nix (I am running [jetpack-nixos](https://github.com/anduril/jetpack-nixos) or adapt to whatever build scripts you are using.

**Comment by conroy-cheers** [comment]
Bump, still pending review

File: vllm/platforms/__init__.py
Comment (youkaichao) [inline vllm/platforms/__init__.py:34]: I need to check with nvidia folks, how robust it is.

File: vllm/platforms/cuda.py
Comment (youkaichao) [inline vllm/platforms/cuda.py:201]: we can have `NvmlCudaPlatform` and `NonNvmlCudaPlatform` inheriting from `Platform`, and in the end of this file, based on jetson or not, define a variable `CudaPlatform` to point to either `NvmlCudaPlatform` or `NonNvmlCudaPlatform` .

**Review by youkaichao — APPROVED** [review]
LGTM now! Thanks for the fix!

**Comment by youkaichao** [comment]
@conroy-cheers can you please merge main to solve the conflict?

**Review by mgoin — APPROVED** [review]
Ditto thanks for the responses!

**Comment by conroy-cheers** [comment]
Merged main and retested on Jetson. Ready to go, assuming tests pass :+1:

File: vllm/platforms/cuda.py
Comment (youkaichao) [inline vllm/platforms/cuda.py:117]: sorry for the additional comments. can you move this function outside of `NvmlCudaPlatform`?

File: vllm/platforms/cuda.py
Comment (youkaichao) [inline vllm/platforms/cuda.py:120]: and call `with nvml_context()` in this function. We don't need to define another `cls._get_physical_device_capability`

Comment (conroy-cheers) [reply vllm/platforms/cuda.py:120]: Should this apply to `_get_physical_device_capability`, `_get_physical_device_name`, and `_get_physical_device_total_memory` as well? thus moving the `@lru_cache(maxsize=8)` and `@with_nvml_context` decorators directly onto the public methods?

Comment (youkaichao) [reply vllm/platforms/cuda.py:120]: yes i think that's even better.

File: vllm/platforms/cuda.py
Comment (conroy-cheers) [reply vllm/platforms/cuda.py]: Done

File: vllm/platforms/cuda.py
Comment (conroy-cheers) [reply vllm/platforms/cuda.py:122]: Done

**Comment by conroy-cheers** [comment]
CI seems to be failing, not sure if it's related to the changes here?

**Comment by mgoin** [comment]
@conroy-cheers failing test is unrelated I'll ask for a force merge, thanks for your work!

**Pull request merged**
